# GDELT API Exploration

## Usage

Note: There is a Python API library open-sourced, which provides handy utilities to interact and query the GDELT Project database. This is [the project's GitHub Page](https://github.com/alex9smith/gdelt-doc-api)


The following is an example to query the API and fetch all articles in the GDELT database containing the one of the keywords ("femicide", "illegal fishing", "wildlife crime") and were published within the timeframe of last 2 months.


In [1]:
from gdeltdoc import GdeltDoc, Filters

In [6]:
f = Filters(keyword=["femicide", "illegal fishing", "wildlife crime"], timespan="2months")
gd = GdeltDoc()

articles = gd.article_search(filters=f)
articles

url  \
0   https://www.elsiglodetorreon.com.mx/noticia/20...   
1   https://www.semana.com/nacion/articulo/urgente...   
2   https://www.eluniversal.com.mx/edomex/caso-ana...   
3   https://www.24-horas.mx/2023/09/17/asesinan-en...   
4   https://www.elsiglodetorreon.com.mx/noticia/20...   
5   https://www.eluniversal.com.co/mundo/el-crudo-...   
6   https://www.eluniversal.com.mx/edomex/caso-ana...   
7   http://www.razon.com.mx/estados/feminicidio-an...   
8   https://tn.com.ar/sociedad/2023/09/17/la-reacc...   
9   https://www.eltiempo.com/bogota/autoridades-re...   
10  https://www.lacapital.com.ar/policiales/un-hom...   
11  https://criteriohidalgo.com/noticias/ana-maria...   
12  https://www.elchubut.com.ar/regionales/2023-9-...   
13  https://www.lapoliticaonline.com:443/politica/...   
14  https://www.laarena.com.ar/la-pampa/la-oposici...   
15  https://finance.sina.com.cn/tech/roll/2023-09-...   
16  https://www.cadena3.com/noticia/cadena-3-elecc...   
17  https://www.cadena3.com/noticia/racing-de-cord...   
18  https://www.cadena3.com/noticia/rony-en-vivo/m...   
19  http://www.catamarcactual.com.ar/nacionales/20...   

                                           url_mobile  \
0   https://www.elsiglodetorreon.com.mx/nota/2023/...   
1   https://www.semana.com/amp/nacion/articulo/urg...   
2   https://www.eluniversal.com.mx/edomex/caso-ana...   
3                                                       
4   https://www.elsiglodetorreon.com.mx/nota/2023/...   
5   https://m.eluniversal.com.co/mundo/el-crudo-re...   
6   https://www.eluniversal.com.mx/edomex/caso-ana...   
7   https://www.razon.com.mx/amp/estados/feminicid...   
8   https://tn.com.ar/sociedad/2023/09/17/la-reacc...   
9   https://www.eltiempo.com/amp/bogota/autoridade...   
10                                                      
11                                                      
12  https://www.elchubut.com.ar/regionales/2023-9-...   
13  https://www.lapoliticaonline.com/amp/269011-la...   
14  https://www.laarena.com.ar/la-pampa/la-oposici...   
15                                                      
16  https://www.cadena3.com/amp/noticia/cadena-3-e...   
17  https://www.cadena3.com/amp/noticia/racing-de-...   
18  https://www.cadena3.com/amp/noticia/rony-en-vi...   
19  https://www.catamarcactual.com.ar/amp/el-radic...   

                                                title          seendate  \
0   Fiscalía de Edomex detiene al presunto feminic...  20230918T031500Z   
1   Capturan en México a exnovio de Ana María Serr...  20230918T030000Z   
2   Caso Ana María : Trasladan al penal de Barrien...  20230918T024500Z   
3   Asesinan en México a Ana María Serrano , sobri...  20230918T033000Z   
4       Sujeto asesina a exnovia en Atizapán , Edomex  20230918T031500Z   
5   El crudo relato de la madre de Ana María Serra...  20230918T033000Z   
6   Caso Ana María : La joven víctima de feminicid...  20230918T024500Z   
7   Feminicidio de Ana María . Así recuerdan a la ...  20230918T033000Z   
8   EL FUERTE MENSAJE de la mamá de CECILIA Strzyz...  20230918T024500Z   
9   Autoridades reportaron cero homicidios durante...  20230918T030000Z   
10  Un hombre en Roldán hirió a balazos a su ex pa...  20230918T033000Z   
11  Ana María , joven de 18 años , fue víctima de ...  20230918T030000Z   
12  De puño y letra : La crisis de Trelew recalent...  20230918T041500Z   
13  La caída de Capitanich : el radicalismo vuelve...  20230918T033000Z   
14  La oposición le arrebató la gobernación a Capi...  20230918T033000Z   
15                     离海最远的地方也产海鲜了 ！ 中国电信新疆公司科技赋能水产业  20230918T041500Z   
16  Río Cuarto : Gonzalo Parodi ganó la interna ra...  20230918T031500Z   
17  Racing de Córdoba jugó el partido como una fin...  20230918T031500Z   
18  Murió uno de los pilotos del avión que se acci...  20230918T031500Z   
19  El radical Zdero derrotó a Capitanich en el Chaco  20230918T024500Z   

                                          socialimage  \
0   https://tecolotit

## 1. Can we use code to fetch lists of URLs matching boolean queries?

Yes, the API query allows for filtering using boolean queries (AND, OR). The below example showcases the example of chaining `OR` keyword and this result is used to query the GDELT API to get a response similar to above.
 
Note: Creating queries using `AND` filter is not directly supported by the library yet and can be achieved with updating some functions and minimal efforts
Note: GDELT API behaves differently when the query string is longer than 225 characters. The current implementation of the API doesn't have a good mechanism to handle such large query strings and the response from GDELT API. The GDELT API in this cases, returns a browser interactive HTML widget. 

In [9]:
f = Filters(keyword=['asesinato', 'homicidio', 'femicidio', 'feminicidio', 'travesticidio', 'transfemicidio', 'Lesbicidio', 'asesina', 'asesinada', 'muerta', 'muerte', 'mata', 'mató', 'dispara', 'balea', 'apuñala', 'acuchillada', 'golpeada', 'estrangula', 'ahogada', 'degollada', 'incinera', 'quemada', 'envenenada', 'prendida fuego', 'descuartizada', 'sin vida', 'intento', 'intento de asesinato', 'Intentó asesinarla', 'intento de femicidio', 'intento de transfemicidio', 'intento de travesticidio', 'intento de lesbicidio', 'intentó matarla', 'suicidio', 'se quito la vida', 'se mató', 'se suicido', 'se ahorco") AND (muje', 'niña', 'una joven', 'una adolescente', 'una chica', 'cuerpo de una mujer', 'restos', 'cadaver de una mujer', 'prostituta', 'trabajadora sexual', 'mujer trans', 'una travesti', 'hombre vestido de mujer'], timespan="2months")
f.query_string

'(asesinato OR homicidio OR femicidio OR feminicidio OR travesticidio OR transfemicidio OR Lesbicidio OR asesina OR asesinada OR muerta OR muerte OR mata OR mató OR dispara OR balea OR apuñala OR acuchillada OR golpeada OR estrangula OR ahogada OR degollada OR incinera OR quemada OR envenenada OR "prendida fuego" OR descuartizada OR "sin vida" OR intento OR "intento de asesinato" OR "Intentó asesinarla" OR "intento de femicidio" OR "intento de transfemicidio" OR "intento de travesticidio" OR "intento de lesbicidio" OR "intentó matarla" OR suicidio OR "se quito la vida" OR "se mató" OR "se suicido" OR "se ahorco") AND (muje" OR niña OR "una joven" OR "una adolescente" OR "una chica" OR "cuerpo de una mujer" OR restos OR "cadaver de una mujer" OR prostituta OR "trabajadora sexual" OR "mujer trans" OR "una travesti" OR "hombre vestido de mujer") &timespan=2months&maxrecords=250'

## 2. If so, how do we filter to results from the last three days and page through those?

Filtering the results based on an absolute and relative time ranges are supported by the GDELT API. For relative filtering this is the supported syntax: 
    - Minutes: `min`
    - Hours: `h`, `hours`
    - Days: `d`, `days`
    - Weeks: `w`, `weeks`
    - Months: `m`, `months` 
    
So, below would be the way to filter the results from last three days

In [10]:
f = Filters(keyword="femicide", timespan="3d")
gd.article_search(f)

url  \
0    https://diario.mx/juarez/encabeza-juarez-top-e...   
1    https://www.am.com.mx/guanajuato/2023/9/17/con...   
2    https://www.eluniversal.com.mx/estados/exigen-...   
3    https://www.radioformula.com.mx/nacional/2023/...   
4    https://www.semana.com/nacion/articulo/aterrad...   
..                                                 ...   
245  https://larepublica.pe:443/sociedad/2023/09/16...   
246  https://www.diariodecuyo.com.ar/argentina/En-C...   
247  https://www.zazoom.it/2023-09-17/violenza-su-d...   
248       https://www.jpost.com/opinion/article-759223   
249  https://laverdadnoticias.com/quintanaroo/Lista...   

                                            url_mobile  \
0    https://diario.mx/juarez/amp/encabeza-juarez-t...   
1    https://www.am.com.mx/amp/congreso-amplia-caus...   
2    https://www.eluniversal.com.mx/estados/exigen-...   
3                                                        
4    https://www.semana.com/amp/nacion/articulo/ate...   
..                                                 ...   
245  https://larepublica.pe/amp/sociedad/2023/09/16...   
246  https://www.diariodecuyo.com.ar/amp/argentina/...   
247                                                      
248         https://m.jpost.com/opinion/article-759223   
249  https://laverdadnoticias.com/amp/quintanaroo/L...   

                                                 title          seendate  \
0        Encabeza Juárez  top  estatal de feminicidios  20230915T200000Z   
1    Congreso amplía causales para aplicar pena máx...  20230917T230000Z   
2                Exigen justicia por maestra asesinada  20230916T104500Z   
3    Encuentran sin vida a Kailani , niña de 3 años...  20230915T201500Z   
4    Aterrador | Ana María Serrano , sobrina del ex...  20230917T220000Z   
..                                                 ...               ...   
245  Abuso infantil : piden aborto terapéutico en Á...  20230916T100000Z   
246  En Chaco ,  cajoneado  el caso Cecilia Strzyzo...  20230917T041500Z   
247  Violenza su donne | Zanella |  castrazione chi...  20230917T143000Z   
248  The shofar call to remember Mahsa Amini - opin...  20230915T081500Z   
249  Lista para ser votada la Ley contra pederastia...  20230916T083000Z   

                                           socialimage                domain  \
0    https://diario.mx/jrz/media/uploads/galeria/20...             diario.mx   
1    https://www.am.com.mx/u/fotografias/m/2023/9/1...             am.com.mx   
2    https://www.eluniversal.com.mx/resizer/CSE8Tf3...    eluniversal.com.mx   
3    https://www.radioformula.com.mx/u/fotografias/...   radioformula.com.mx   
4    https://www.semana.com/resizer/cH2C1EzR8Xmeez0...            semana.com   
..                                                 ...                   ...   
245  https://imgmedia.larepublica.pe/1200x630/larep...        larepublica.pe   
246  https://www.diariodecuyo.com.ar/__export/16949...   diariodecuyo.com.ar   
247                                                                zazoom.it   
248  https://images.jpost.com/image/upload/f_auto,f...             jpost.com   
249  https://laverdadnoticias.com/__export/16948181...  laverdadnoticias.com   

    language sourcecountry  
0    Spanish        Mexico  
1    Spanish        Mexico  
2    Spanish        Mexico  
3    Spanish        Mexico  
4    Spanish      Colombia  
..       ...           ...  
245  Spanish          Peru  
246  Spanish     Argentina  
247  Italian         Italy  
248  English        Israel  
249  Spanish        Mexico  

[250 rows x 8 columns]

Note: Currently there is no mechanism to paginate or retrieve more than 250 articles in a single API call. None of the online resources or blogs on GDELT website specify alternative mechanisms to achieve this. But, this should be relatively easy to achieve, through the use of multiple API calls and filtering the results for each day or week and concatenating the result sets.  

## 3. Can we filter by geographic country or state/provide of publication?

Yes, the results of the search can be filtered based on the source country of publication, along with the language of publication. The API uses the standard [FIPS country code standard](https://en.wikipedia.org/wiki/List_of_FIPS_country_codes) to filter based on the country. There 

In [11]:
f = Filters(keyword=["femicide", "farmer"], timespan="2m", country=["IN", "PK"])
gd.article_search(f)

url  \
0    https://www.thehindu.com/news/international/tu...   
1    https://tribune.com.pk/story/2429155/corporate...   
2                     https://jang.com.pk/news/1263261   
3    https://dunyanews.tv/en/Entertainment/749835-F...   
4                     https://jang.com.pk/news/1256337   
..                                                 ...   
245  https://www.punjabitribuneonline.com/news/doab...   
246  https://www.amarujala.com/dehradun/the-farmers...   
247  https://www.punjabitribuneonline.com/news/punj...   
248  https://www.kannadaprabha.com/karnataka/2023/a...   
249  https://www.sangbadpratidin.in/offbeat/tamil-n...   

                                            url_mobile  \
0    https://www.thehindu.com/news/international/tu...   
1                                                        
2                      https://jang.com.pk/amp/1263261   
3                                                        
4                      https://jang.com.pk/amp/1256337   
..                                                 ...   
245  https://m.punjabitribuneonline.com/article/the...   
246  https://www.amarujala.com/amp/dehradun/the-far...   
247  https://m.punjabitribuneonline.com/article/lon...   
248  https://m.kannadaprabha.com/karnataka/2023/aug...   
249  https://m.sangbadpratidin.in/article/tamil-nad...   

                                                 title          seendate  \
0    Turkey drops bid to close leading women rights...  20230914T003000Z   
1           Corporate farming to attract Rs100 billion  20230804T083000Z   
2    بھارت : کسانوں کا احتجاج ، مظاہرین عمارت کے حف...  20230829T171500Z   
3             Farmer stumbles upon big  dinosaur egg    20230824T223000Z   
4    جھل مگسی میں کسان کے قتل کا نو ٹس لیا جا ئے ، ...  20230810T033000Z   
..                                                 ...               ...   
245  ਕਿਸਾਨਾਂ - ਮਜ਼ਦੂਰਾਂ ਨੇ ਕੇਂਦਰ ਸਰਕਾਰ ਦੇ ਪੁਤਲੇ ਫੂਕ...  20230909T061500Z   
246  Dehradun News : बीज की कमी के कारण चिंतित नजर ...  20230723T203000Z   
247  ਲੌਂਗੋਵਾਲ : ਪੁਲੀਸ ਨਾਲ ਝੜਪ ਦੌਰਾਨ ਕਿਸਾਨ ਦੀ ਮੌਤ - ...  20230822T034500Z   
248  ತಿ . ನರಸಿಪುರ ರಸ್ತೆ ಅಪಘಾತದಲ್ಲಿ ರೈತ ಸಾವು : ಹೆದ್ದ...  20230826T114500Z   
249  Tamil Nadu Couple Act Fake Accident and Hijack...  20230723T111500Z   

                                           socialimage  \
0    https://th-i.thgim.com/public/incoming/hktweg/...   
1    https://i.tribune.com.pk/media/images/JSSKHSJR...   
2    https://jang.com.pk/assets/uploads/updates/202...   
3    https://img.dunyanews.tv/news/2023/August/08-2...   
4                                                        
..                                                 ...   
245  https://www.punjabitribuneonline.com/wp-conten...   
246  https://staticimg.amarujala.com/assets/images/...   
247  https://www.punjabitribuneonline.com/wp-conten...   
248  https://media.kannadaprabha.com/uploads/user/i...   
249  https://www.sangbadpratidin.in/wp-content/uplo...   

                       domain language sourcecountry  
0                thehindu.com  English         India  
1              tribune.com.pk  English      Pakistan  
2                 jang.com.pk     Urdu      Pakistan  
3                dunyanews.tv  English      Pakistan  
4                 jang.com.pk     Urdu      Pakistan  
..                        ...      ...           ...  
245  punjabitribuneonline.com  Punjabi         India  
246             amarujala.com    Hindi         India  
247  punjabitribuneonline.com  Punjabi         India  
248         kannadaprabha.com  Kannada         India  
249        sangbadpratidin.in  Bengali         India  

[250 rows x 8 columns]

## 4. Are there API limits? Are there any associated monetary costs?

There doesn't seem to be a monetary or pricing mechanism imposed on the GDELT API. Their website describes themselves as 
```
The entire GDELT database is 100% free and open and you can download the raw datafiles, visualize it using the
GDELT Analysis Service, or analyze it at limitless scale with Google BigQuery.
```
So, there shouldn't be any associated monetary costs without any information from their blogs.


The following are the limitations that I have noticed during my interaction with GDELT: 
- Max 250 articles are returned when querying through API
- Doesn't have support for fuzzy matching text within articles. So, the filters would have to account for language tense, sentence structure, etc while designing.


## Conclusion

While working with the GDELT API, I found that it supports a few other features as well:
- Allows for searching images (attached to articles) using the image EXIF data, image OCR result, image tag (images are processed by GDELT and tagged based on the image content predictions)
- Allows for filtering based on the tonality of articles (how negative or positive the article's content)
- Allows for filtering based on themes (pre-determined set of categories, examples: TERROR, CRISIS, MARITIME, etc). List of themes [is here](http://data.gdeltproject.org/api/v2/guides/LOOKUP-GKGTHEMES.TXT).
- It can return visualizations for certain contexts based on the query. (This can save the processing overhead for certain situations)


## References
- GDELT API Documentation: https://blog.gdeltproject.org/gdelt-doc-2-0-api-debuts/
- GDELT Python API package: https://github.com/alex9smith/gdelt-doc-api
- GDELT Blog (for extra info): https://blog.gdeltproject.org/